adcc/adc_ea/matrix.py

```diff
@@ -0,0 +1,232 @@
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with adcc. If not, see <http://www.gnu.org/licenses/>.
+##
+## ---------------------------------------------------------------------
+from math import sqrt
+from collections import namedtuple
+
+from adcc import block as b
+from adcc.functions import direct_sum, einsum
+from adcc.Intermediates import Intermediates, register_as_intermediate
+from adcc.AmplitudeVector import AmplitudeVector
+
+__all__ = ["block"]
+
+# TODO One thing one could still do to improve timings is implement a "fast einsum"
+#      that does not call opt_einsum, but directly dispatches to libadcc. This could
+#      lower the call overhead in the applies for the cases where we have only a
+#      trivial einsum to do. For the moment I'm not convinced that is worth the
+#      effort ... I suppose it only makes a difference for the cheaper ADC variants
+#      (ADC(0), ADC(1), CVS-ADC(0-2)-x), but then on the other hand they are not
+#      really so much our focus.
+
+
+#
+# Dispatch routine
+#
+"""
+`apply` is a function mapping an AmplitudeVector to the contribution of this
+block to the result of applying the ADC matrix. `diagonal` is an
+`AmplitudeVector` containing the expression to the diagonal of the ADC matrix
+from this block.
+"""
+AdcBlock = namedtuple("AdcBlock", ["apply", "diagonal"])
+
+
+def block(ground_state, spaces, order, variant=None, intermediates=None):
+    """
+    Gets ground state, potentially intermediates, spaces (ph, pphh and so on)
+    and the perturbation theory order for the block,
+    variant is "cvs" or sth like that.
+
+    It is assumed largely, that CVS is equivalent to
+    mp.has_core_occupied_space, while one would probably want in the long run
+    that one can have an "o2" space, but not do CVS.
+    """
+    if isinstance(variant, str):
+        variant = [variant]
+    elif variant is None:
+        variant = []
+    reference_state = ground_state.reference_state
+    if intermediates is None:
+        intermediates = Intermediates(ground_state)
+
+    if ground_state.has_core_occupied_space and "cvs" not in variant:
+        raise ValueError("Cannot run a general (non-core-valence approximated)"
+                         " ADC method on top of a ground state with a "
+                         "core-valence separation.")
+    if not ground_state.has_core_occupied_space and "cvs" in variant:
+        raise ValueError("Cannot run a core-valence approximated ADC method on"
+                         " top of a ground state without a "
+                         "core-valence separation.")
+
+    fn = "_".join(["block"] + variant + spaces + [str(order)])
+
+    if fn not in globals():
+        raise ValueError("Could not dispatch: "
+                         f"spaces={spaces} order={order} variant=variant")
+    return globals()[fn](reference_state, ground_state, intermediates)
```

Comment:
Except for the `globals()` stuff, this is identical to the other matrix functions, right? I wonder if there's a way to reduce code duplication...

Reply:
What could be done is adding a new function in adc_pp/matrix.py or maybe in a new general file:
```
def get_block_prereqs(ground_state, spaces, order, variant=None,
                        intermediates=None):
    if isinstance(variant, str):
        variant = [variant]
    elif variant is None:
        variant = []
    reference_state = ground_state.reference_state
    if intermediates is None:
        intermediates = Intermediates(ground_state)

    if ground_state.has_core_occupied_space and "cvs" not in variant:
        raise ValueError("Cannot run a general (non-core-valence approximated) "
                         "ADC method on top of a ground state with a "
                         "core-valence separation.")
    if not ground_state.has_core_occupied_space and "cvs" in variant:
        raise ValueError("Cannot run a core-valence approximated ADC method on "
                         "top of a ground state without a "
                         "core-valence separation.")

    return ("_".join(["block"] + variant + spaces + [str(order)]),
            reference_state)
```

and then importing it in the ip and ea matrix.py files and using the block routine as:

```
def block(ground_state, spaces, order, variant=None, intermediates=None):
    fn, reference_state = get_block_prereqs(ground_state, spaces, order,
                                       variant=None, intermediates=None)

    if fn not in globals():
        raise ValueError("Could not dispatch: "
                         f"spaces={spaces} order={order} variant=variant")
    return globals()[fn](reference_state, ground_state, intermediates)
```
    
I think more cannot be reduced while keeping this structure. Also the 
`AdcBlock = namedtuple("AdcBlock", ["apply", "diagonal"])`
seems necessary.
I'm not sure about the `_all_=` line since I think wildcard imports don't occur for the matrices. So if it is needed in general?